priv/post.ex.eex

```diff
@@ -30,22 +29,21 @@ defmodule <%= context.module_name %> do
     ]
 <%= if context.protocol == "query" do %>
     input = Map.merge(input, %{"Action" => action, "Version" => "<%= context.api_version %>"})<% end %>
-    payload = <%= context.encode %>
+    payload = encode!(client, input)
     headers = AWS.Request.sign_v4(client, "POST", url, headers, payload)
+    post(client, url, payload, headers, options)
+  end
 
-    case HTTPoison.post(url, payload, headers, options) do
-      {:ok, %HTTPoison.Response{status_code: 200, body: ""} = response} ->
-        {:ok, nil, response}
-
-      {:ok, %HTTPoison.Response{status_code: 200, body: body} = response} ->
-        {:ok, <%= context.decode %>, response}
+  defp post(client, url, payload, headers, options) do
+    case do_request(client, :post, url, payload, headers, options) do
+      {:ok, %{status_code: 200, body: body} = response} ->
+        body = if(body != "", do: decode!(client, body))
+        {:ok, body, response}
 
-      {:ok, %HTTPoison.Response{body: body}} ->
-        error = <%= context.decode %>
-        {:error, error}
+      {:ok, %{body: body}} ->
+        {:error, decode!(client, body)}
```

Comment:
We return the raw response as a third argument when successful.  Should we return it here too? 

Like this:
```elixir
{:error, %{reason: decode!(client, body), response: response}}
```

Reply:
I think it's fair to return the response, but I would suggest a tuple with three elements, like we do on success:

```elixir
{:error, decode!(client, body), response}
```

Another option is to introduce a custom response struct. This way we could only return a `{:ok, response} | {:error, response}`, and have the body as the decoded version. We could introduce an option to keep the original body in a `raw_body` field.

@josevalim @vrcca what do you think?

Reply:
I think this function is internal (not a public contract), so we probably don't need to be consistent here and just use what we actually need. If this was somewhat public API through or part of the HTTPClient/JSON contracts, I would agree.

Reply:
Checking the generated code, I believe this is a public contract @josevalim:
POST example
https://github.com/aws-beam/aws-elixir/blob/master/lib/aws/sns.ex#L491

REST example
https://github.com/aws-beam/aws-elixir/blob/master/lib/aws/cloud_directory.ex#L1059

Reply:
As you prefer then! :)

Reply:
@philss If we change this, we will break the public API. Do you think it's worth it?

Reply:
We can break it if we want to as it is before 1.0.

Reply:
Nice! Then wdyt we go with a simple 3 element tuple for now, and improve this later? @philss @josevalim

Reply:
Actually I think the {:error, response} is a bit confusing so we should change it. We also don’t always have a response, so we need to take that into account. Maybe we should just return {:error, {:invalid_response, whole_response_with_status}}?

Reply:
I like José's suggestion. It will add more flexibility to the client.

Reply:
I like it too. Although I named it `unexpected_response`, because they may be valid responses, but unexpected returns 🤔